### ISSUE TITLE
Unable to run this_works_1_18 notebook with fresh / clean finrl installation

### ISSUE THREAD

**Issue by RajS999** [AI4Finance-Foundation/FinRL#167]
To keep it as clean as possible, I uninstalled and reinstalled python 3.7.7 in Windows. Then, I created new pipenv environment and installed only finrl library with command `pipenv install finrl`. Then to avoid any possible dependency / package version issues that installing jupyter may cause, I exported [this_works_1_18 notebook](https://github.com/AI4Finance-LLC/FinRL-Library/blob/master/notebooks/this_works_1_18.ipynb) as py file and then debugged in vscode. 

When I step over this line:

```
df_account_value, df_actions = DRLAgent.DRL_prediction(model=model,environment = e_trade_gym)
```

it printed this line in the console:

```
PISODE|STEPS|TERMINAL_REASON|TOT_ASSETS|TERMINAL_REWARD_unsc|CASH_PCT
   1| 101|CASH SHORTAGE  |$1011236  |-0.18616% |0.30%
hit end!
```

and then gave following error in the debugger:

```
Traceback (most recent call last):
  <vscode debugger related stacktrace>
  File "f:\RajS999\workspaces\stockpipgit\this_works_1_18_original.py", line 790, in <module>
    df_account_value, df_actions = DRLAgent.DRL_prediction(model=model,environment = e_trade_gym)
  File "C:\Users\RajS\.virtualenvs\stockpipgit-_flxlVEa\lib\site-packages\finrl\model\models.py", line 88, in DRL_prediction
    return account_memory[0], actions_memory[0]
IndexError: list index out of range
```

Before starting debugging, I have made following changes to this py file:

- Commented out `from finrl.env.env_stocktrading_v2 import StockTradingEnvV2`, since import fails as `StockTradingEnvV2` is not there in the main branch (this class is explicitly included in the notebook)
- Commented `env_train, _ = e_train_gym.get_multiproc_env(n = n_cores)` and uncommented `env_train, _ = e_train_gym.get_sb_env()` since I was getting this [error](https://github.com/lyft/l5kit/issues/129) which seems to be possible know bug.
- I reduced learn steps from `model.learn(total_timesteps = 5000000, ...` to `model.learn(total_timesteps = 2000, ...`, to avoid waiting for long (as I was just trying to quickly run the notebook).

Is the last change culprit?

Before all above efforts, I tried to run this notebook in jupyter. This is is google colab screenshot of the error:

![image](https://user-images.githubusercontent.com/52235655/110181373-a7e17500-7e31-11eb-9efd-9dcc619d3de1.png)

I was am not able to run other notebooks also. Then while going through issues, I realised that single stock notebook is merged into multistock and I should be trying out latest notebook which is supposedly this one, right? Or will this work in docker? Havent tried docker route yet. Please give some pointers...

**PS:**

I also tried running same py file in clean pipenv environment with finrl intalled with the command `pipenv install git+https://githu...`, but getting same error.

**Comment by spencerR1992** [comment]
@RajS999 please try the env_multistock cashpenalty notebook.

Ultimately, the best way to guarantee that the code works is to install and run with Docker. We simply cannot support all environment types.

**Comment by RajS999** [comment]
@spencerR1992 increasing `total_timesteps` in `model.learn(total_timesteps = 2000, ...` to `model.learn(total_timesteps = 20000, ...` seem to have resolved it, though I did not get exactly why ?

**Comment by s-diaco** [comment]
I have the same issue. I've tried it in docker and tried "env_multistock_cashpenalty.pynb" notebook as suggested. Also I set total_timesteps = 2000 and 20000. Same error as op:

`IndexError: list index out of range`

when stepping over this line:

`df_account_value, df_actions = DRLAgent.DRL_prediction(model=model,environment = e_trade_gym)`

**Comment by s-diaco** [comment]
I found a workaround for the problem (works in "env_multistock_cashpenalty.pynb" notebook):
If you add

`e_trade_gym.patient = True`

you will get no Error

before adding the line:
![before](https://user-images.githubusercontent.com/17494033/111055384-92e88f80-848a-11eb-8b95-60585333cce7.jpg)

after adding the line:
![after](https://user-images.githubusercontent.com/17494033/111055380-8fed9f00-848a-11eb-8551-ea50bab64ae8.jpg)

The problem:
This is method "DRL_prediction" in "models.py" file:

```
    @staticmethod
    def DRL_prediction(model, environment):
        test_env, test_obs = environment.get_sb_env()
        """make a prediction"""
        account_memory = []
        actions_memory = []
        test_env.reset()
        for i in range(len(environment.df.index.unique())):
            action, _states = model.predict(test_obs)
            #account_memory = test_env.env_method(method_name="save_asset_memory")
            #actions_memory = test_env.env_method(method_name="save_action_memory")
            test_obs, rewards, dones, info = test_env.step(action)
            if i == (len(environment.df.index.unique()) - 2):
              account_memory = test_env.env_method(method_name="save_asset_memory")
              actions_memory = test_env.env_method(method_name="save_action_memory")
            if dones[0]:
                print("hit end!")
                break
        return account_memory[0], actions_memory[0]
```

If cash becomes zero or less in the middle of the prediction, the method enters the secound "if" clause without ever entering the first "if" clause that assigns the return values. So it goes to the last line:

`return account_memory[0], actions_memory[0]`

while both return values ("account_memory" and "actions_memory") are empty arrays and have nothing to return.

**Comment by BruceYanghy** [comment]
Only need to call account_memory and actions_memory once in the last time step, don’t need to put that in the loop.

**Comment by Mahesha999** [comment]
Just to add, reducing `hmax` also eliminates error. 
Notice in first pic, hmax is set to 500 in first line and in second its 1000 which gives error:

`hmax=500`:

![image](https://user-images.githubusercontent.com/4706946/113504204-4574b500-9554-11eb-9a45-8e3ce2254a6a.png)

`hmax=1000`:

![image](https://user-images.githubusercontent.com/4706946/113504218-57565800-9554-11eb-91a1-d2ebbfb1f5d9.png)

PS: This is [cash_penalty](https://github.com/AI4Finance-LLC/FinRL-Library/blob/master/notebooks/env_multistock_cashpenalty.ipynb) notebook with only difference that I trained it for just 30000 timesteps (`model.learn(total_timesteps=30000, .... )`).

**Comment by CITIZENDOT** [comment]
Thanks @Mahesha999

**Comment by timweissenfels** [comment]
Why does a higher hmax cause the error? Has this something to do with hardware requirements? @rayrui312
